streams/src/main/java/org/apache/kafka/streams/state/internals/MemoryLRUCache.java

```diff
@@ -105,14 +100,28 @@ public boolean persistent() {
     }
 
     @Override
-    public V get(K key) {
-        return this.map.get(key);
+    public boolean open() {
+        return open;
+    }
+
+    @Override
+    public  V get(K key) {
+        lock.readLock().lock();
+        try {
+            return this.map.get(key);
+        } finally {
+            lock.readLock().unlock();
+        }
     }
 
     @Override
     public void put(K key, V value) {
-        this.map.put(key, value);
-        this.keys.add(key);
+        lock.writeLock().lock();
```

Comment:
Thanks for the review and information @ben-manes. We don't yet require Java 8 and hence why Caffeine is probably not an option for now, but definitely something to consider for the future. I agree that simply using synchronized for now is better than the read write lock.